sdk/python/kfp_tekton/tekton.py

```diff
@@ -311,3 +311,39 @@ def __enter__(self) -> Union[Tuple[TektonLoopIterationNumber, LoopArguments], _f
   def enumerate(self) -> dsl.ParallelFor:
     self.call_enumerate = True
     return self
+
+
+class AddOnGroup(dsl.OpsGroup):
+    """
+    Represents an AddOn group containing ops and group of OpsGroups.
+    This class is the base class for a customized OpsGroup. Users can
+    develop their own OpsGroup. The customized OpsGroup maps to a
+    custom task in Tekton.
+    """
+    TYPE_NAME = 'addon_group'
+    TASK_TYPE = 'task'
+    RUN_TYPE = 'run'  # means custom task
+    DEFAULT_KIND = 'AddOnGroup'
+    DEFAULT_APIVERSION = 'custom.tekton.dev/v1alpha1'
+
+    def __init__(self, task_type: str = RUN_TYPE,
+                kind: str = DEFAULT_KIND,
+                api_version: str = DEFAULT_APIVERSION,
+                is_finally: bool = False,
```

Comment:
I don't think that's how we used to generate `finally` in kfp-tekton for usual tasks. I'm not sure if the DSL users will be ok with that...

I would expect to be able to write sth like:
```python
exit_op = Handler()
with exit_op:
  exit_handler_op_0 = ...
  exit_handler_op_1 = ...

with dsl.ExitHandler(exit_op):
  main_op_0 = ...
  main_op_1 = ...
```
which would produce sth like:
```yaml
tasks:
- name: main-op-0
  ...
- name: main-op-1
  ...
finally:
- name: addon-group-1
  params: ...
  taskSpec:
    apiVersion: custom.tekton.dev/v1alpha1
    kind: Handler
    spec:
      pipelineSpec:
        params: ...
        tasks:
        - name: exc-handler-op-0
          ...
        - name: exc-handler-op-1
          ...
```

...but it doesn't seem to work, failing with `AttributeError: 'Handler' object has no attribute 'dependent_names'`. When I add that fields in the constructor, the following is produced:
```yaml
tasks:
- name: main-op-0
  ...
- name: main-op-1
  ...
- name: addon-group-1
  params: ...
  taskSpec:
    apiVersion: custom.tekton.dev/v1alpha1
    kind: Exception
    spec:
      pipelineSpec:
        params: ...
        tasks:
        - name: exc-handler-op-0
          ...
        - name: exc-handler-op-1
          ...
```

Reply:
When adding `AddOnGroup`, I made it be an `OpsGroup`, which means it can be used to mimic `ExitHandler`. For `ExitHandler`, its constructor tasks a `BaseOp`, but not` OpsGroup`.

Here is an example of using `is_finally=True`:
```python
class Exception(AddOnGroup):
  """Exception Handler"""

  def __init__(self, params: Dict[str, Union[dsl.PipelineParam, str, int]] = {}):
    params['pipelinerun_name']= '$(context.pipelineRun.name)'
    super().__init__(kind='Exception', params=params, api_version='custom.tekton.dev/v1alpha1', is_finally=True)


@dsl.pipeline(
    name='exception-handler',
    description='Addon sample'
)
def download_and_print(url: str = 'gs://ml-pipeline-playground/shakespeare1.txt'):
    """A sample pipeline showing exit handler."""

    echo_op('echo task!')

    with Exception():
        task1()
        task2()
```

and the output YAML should contain the finally section like this:
```yaml
    finally:
    - name: addon-group-1
      params:
      - name: pipelinerun_name
        value: $(context.pipelineRun.name)
      taskSpec:
        apiVersion: custom.tekton.dev/v1alpha1
        kind: Exception
        spec:
          pipelineSpec:
            params:
            - name: pipelinerun_name
              type: string
            tasks:
            - name: task1
              ...
            - name: task2
              ....
```

Reply:
Hmm... I see. Considering how we always have only one `finally` task anyway, it sounds ok.